File: problems/2047_number_of_valid_words_in_a_sentence/solution.py
class Solution:
    def countValidWords(self, sentence: str) -> int:
        def isValid(token: str) -> bool:
            hyphen = 0
            for i, c in enumerate(token):
                if c.isdigit():
                    return False
                if c == '-':
                    hyphen += 1
                    if hyphen > 1:
                        return False
                    if i == 0 or i == len(token) - 1:
                        return False
                    if not token[i - 1].isalpha() or not token[i + 1].isalpha():
                        return False
                if c in '!.,;':
                    if i != len(token) - 1:
                        return False
            return len(token) > 0
        return sum(isValid(t) for t in sentence.split())

File: problems/2047_number_of_valid_words_in_a_sentence/Solution.java
class Solution {
    public int countValidWords(String sentence) {
        int count = 0;
        for (String token : sentence.split("\\s+")) {
            if (token.isEmpty()) continue;
            if (isValid(token)) count++;
        }
        return count;
    }
    private boolean isValid(String token) {
        int hyphen = 0;
        for (int i = 0; i < token.length(); i++) {
            char c = token.charAt(i);
            if (Character.isDigit(c)) return false;
            if (c == '-') {
                hyphen++;
                if (hyphen > 1) return false;
                if (i == 0 || i == token.length() - 1) return false;
                if (!Character.isLetter(token.charAt(i - 1)) || !Character.isLetter(token.charAt(i + 1))) return false;
            }
            if ("!.,;".indexOf(c) >= 0 && i != token.length() - 1) return false;
        }
        return true;
    }
}
